WIP: rebuild qmck in cpp

tfki opened the pull request:
- implement operator<< for extended_logic_bundle
 - add known correct example from uni for verification purposes
 - work on qmck::deduce
 - fix a couple minor things

Commit:
WIP: rebuild qmck in cpp
 - implement operator<< for extended_logic_bundle
 - add known correct example from uni for verification purposes
 - work on qmck::deduce
 - fix a couple minor things

rawbby commented on core/qmckcpp/include/qmck/extended_logic_table.hpp:
```diff
@@ -25,8 +25,12 @@ namespace qmck
 
         extended_logic_table(extended_logic_table &&lhs) noexcept = default;
 
+        extended_logic_table(logic_table_format &format);
```
please use a const reference here

rawbby commented on core/qmckcpp/src/qmck.cpp:
```diff
@@ -64,29 +65,29 @@ namespace qmck
         //    +-> modify the output mask from 1. reduction_logic_table for found merges
 
-        for (int x = 0; x < 32; ++x)
+        std::list<extended_logic_table> all_extended_logic_tables;
```
better use the std::vector instead of a std::list (performance matters)

rawbby commented on core/qmckcpp/src/qmck/extended_logic_table.cpp:
```diff
@@ -35,11 +39,16 @@ qmck::extended_logic_table::extended_logic_table(qmck::logic_table &&lhs)
     }
 }
 
+qmck::extended_logic_table::extended_logic_table(qmck::logic_table_format &format)
```
you may want to also implement a move construction for this. In case the format was created only for this object...

rawbby commented on core/qmckcpp/src/qmck/extended_logic_table.cpp:
```diff
@@ -52,3 +61,15 @@ void qmck::swap(qmck::extended_logic_table &first, qmck::extended_logic_table &s
+
+bool qmck::extended_logic_table::empty()
+{
+    for (auto &bundle:logic_bundle_ranks)
```
this should be an "auto const &"

rawbby commented on core/qmckcpp/src/qmck/extended_logic_table.cpp:
```diff
@@ -52,3 +61,15 @@ void qmck::swap(qmck::extended_logic_table &first, qmck::extended_logic_table &s
     swap(first.logic_bundle_ranks, second.logic_bundle_ranks);
 }
+
+bool qmck::extended_logic_table::empty()
```
this method does not modify any members, therefor it should be declared const. (void foo empty() const { ... } )

rawbby commented on core/qmckcpp/src/qmck.cpp:
```diff
@@ -64,29 +65,29 @@ namespace qmck
                 for (int lower_i{0}; lower_i < lower_n; lower_i++)
                 {
                     auto upper_n = upper_bundle.inputs.size();
                     for (int upper_i{0}; upper_i < upper_n; upper_i++)
```
use size_t here

rawbby commented on core/qmckcpp/src/qmck.cpp:
```diff
@@ -97,27 +98,43 @@ namespace qmck
+        // remove empty one at the end
+        all_extended_logic_tables.pop_back();
+
+        int counter{0};
```
use size_t here as the counter is used in context of a vector iteration

rawbby commented on core/qmckcpp/src/qmck/extended_logic_table.cpp:
```diff
@@ -4,10 +4,12 @@
-        if (__builtin_popcount(lhs.outputs[i])) // ignore zero output
+        if (lhs.outputs[i]) // ignore zero output
         {
             uint8_t rank = __builtin_popcount(lhs.inputs[i]);
```
use the "auto + static_cast" idiom here to show that here is a wanted conversion. ( auto rank = static_cast<uint8_t>(__builtin_popcount(lhs.inputs[i])); )

rawbby commented on core/qmckcpp/src/qmck/extended_logic_table.cpp:
```diff
@@ -16,17 +18,19 @@ qmck::extended_logic_table::extended_logic_table(qmck::logic_table const &lhs)
-        if (__builtin_popcount(lhs.outputs[i])) // ignore zero output
+        if (lhs.outputs[i]) // ignore zero output
         {
             uint8_t rank = __builtin_popcount(lhs.inputs[i]);
```
use the "auto + static_cast" idiom here to show that here is a wanted conversion. ( auto rank = static_cast<uint8_t>(__builtin_popcount(lhs.inputs[i])); )

rawbby commented on core/qmckcpp/src/qmck_io.cpp:
```diff
@@ -25,6 +25,58 @@ std::ostream &operator<<(std::ostream &out, const qmck::extended_logic_bundle &l
+    out << "logic_table_format=" << lhs.format << ',' << std::endl << std::endl;
+    out << "----------------" << std::endl;
+
+    for (int bundle_i{0}; bundle_i < lhs.logic_bundle_ranks.size(); bundle_i++)
```
use size_t here

rawbby commented on core/qmckcpp/src/qmck_io.cpp:
```diff
@@ -25,6 +25,58 @@ std::ostream &operator<<(std::ostream &out, const qmck::extended_logic_bundle &l
+
+    for (int bundle_i{0}; bundle_i < lhs.logic_bundle_ranks.size(); bundle_i++)
+    {
+        auto &current_bundle = lhs.logic_bundle_ranks[bundle_i];
```
use "auto const &" here as the object is not modified

rawbby commented on core/qmckcpp/src/qmck_io.cpp:
```diff
@@ -25,6 +25,58 @@ std::ostream &operator<<(std::ostream &out, const qmck::extended_logic_bundle &l
+        auto &current_bundle = lhs.logic_bundle_ranks[bundle_i];
+        if (!current_bundle.inputs.empty())
+        {
+            for (int row_i{0}; row_i < current_bundle.inputs.size(); row_i++)
```
use size_t here

rawbby commented on core/qmckcpp/src/qmck_io.cpp:
```diff
@@ -25,6 +25,58 @@ std::ostream &operator<<(std::ostream &out, const qmck::extended_logic_bundle &l
+            for (int row_i{0}; row_i < current_bundle.inputs.size(); row_i++)
+            {
+
+                for (int j = lhs.format.inputs_count - 1; j >= 0; j--)
```
use auto here

rawbby commented on core/qmckcpp/src/qmck_io.cpp:
```diff
@@ -25,6 +25,58 @@ std::ostream &operator<<(std::ostream &out, const qmck::extended_logic_bundle &l
+                }
+                out << " | ";
+
+                for (int k = lhs.format.outputs_count - 1; k >= 0; k--)
```
use auto here

rawbby commented on core/qmckcpp/src/qmck_io.cpp:
```diff
@@ -25,6 +25,58 @@ std::ostream &operator<<(std::ostream &out, const qmck::extended_logic_bundle &l
+                }
+                out << "  ";
+
+                for (int l = lhs.format.outputs_count - 1; l >= 0; l--)
```
use auto here

rawbby commented on core/qmckcpp/src/qmck_io.cpp:
```diff
@@ -36,11 +88,10 @@ std::ostream &operator<<(std::ostream &out, const qmck::logic_array &lhs)
-    out << '\t' << "logic_table_format=" << lhs.format << ',' << std::endl << std::endl;
+    out << "logic_table_format=" << lhs.format << ',' << std::endl << std::endl;
 
     for (int i = 0; i < lhs.inputs.size(); i++)
```
use size_t here

rawbby commented on core/qmckcpp/src/qmck_io.cpp:
```diff
@@ -36,11 +88,10 @@ std::ostream &operator<<(std::ostream &out, const qmck::logic_array &lhs)
     for (int i = 0; i < lhs.inputs.size(); i++)
     {
-        out << '\t';
         for (int j = lhs.format.inputs_count - 1; j >= 0; j--)
```
use size_t here

rawbby commented on core/qmckcpp/include/qmck/extended_logic_table.hpp:
```diff
@@ -25,8 +25,12 @@ namespace qmck
+
     public:
 
+        bool empty();
```
declare const here, as this is a read only method

rawbby commented on core/qmckcpp/src/qmck.cpp:
```diff
@@ -1,12 +1,13 @@
     qmck::extended_logic_table deduce(qmck::logic_table table)
     {
-        std::cout << table;
+        std::cout << table << std::endl;
```
use '\n' if you want better performance
as std::endl uses a heavy evaluation

rawbby commented on core/qmckcpp/src/qmck.cpp:
```diff
@@ -97,27 +98,43 @@ namespace qmck
+        all_extended_logic_tables.pop_back();
+
+        int counter{0};
+        for (auto &extended_logic_table:all_extended_logic_tables)
```
use "auto const &" here

rawbby commented on core/qmckcpp/src/qmck/extended_logic_table.cpp:
```diff
@@ -16,17 +18,19 @@ qmck::extended_logic_table::extended_logic_table(qmck::logic_table const &lhs)
+    this->format = lhs.format;
+
     auto n = lhs.inputs.size();
     for (size_t i = 0; i < n; i++)
```
to ensure best performance over all compilers use ++i here.

Commit:
fix issues mentioned in pull-request

Commit:
tune deduce(), fix bug
- save only prime rows and not all reduction tables, as this is much more memory efficient

Commit:
delete duplicate rows in qmck::deduce result

Commit:
refactor qmck::extended_logic_table and qmck::extended_logic_bundle
for much easier/shorter usage and memory optimization

Commit:
never use uintX_t when working with qmck::logic_value
so the logic_value typedef can esasily be changed,
same applies to the size of qmck::quine_table::ranks size

Commit:
add row indices to printed quine_table, fix sample input files

Commit:
fix input file format
now all rows require a ';' terminating them

Commit:
refactoring
replace usage of size_t with std::size_t since only c headers are guaranteed to declare size_t in global namespace

Commit:
cleanup & add support to build tree from qmck result

Commit:
solve merge conflicts, replace size_t with std::size_t in some places

Commit:
refactor tree structure because former was impractical
